File: src/main/java/guru/xgm/jass/openapi/actionSystem/convert/jass2lua/Jass2LuaVisitor.java
package guru.xgm.jass.openapi.actionSystem.convert.jass2lua;

import guru.xgm.jass.openapi.actionSystem.convert.Jass2AnyVisitor;
import guru.xgm.jass.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import static guru.xgm.jass.psi.JassTypes.*;

public class Jass2LuaVisitor extends Jass2AnyVisitor {
    @Override
    public void appendStatementLineEnd() {
        stringBuffer.append("\n");
    }

    @Override
    public void appendSingleLineComment(String comment) {
        stringBuffer.append("--").append(comment).append("\n");
    }

    @Override
    public void appendBlockLineComment(String comment) {
        stringBuffer.append("--[[").append(comment).append("--]]");
    }

    @Override
    public void appendNull() {
        stringBuffer.append("nil");
    }

    // -- variable
    @Override
    public void appendVar(boolean constant, boolean global, boolean array, @NotNull String type, String name, @Nullable JassExpr expr) {
        stringBuffer.append("---@type ").append(type);
        if (global) stringBuffer.append(" @global");
        if (constant) stringBuffer.append(" constant");
        stringBuffer.append("\n");

        if (!global) stringBuffer.append("local ");
        stringBuffer.append(name);

        stringBuffer.append(" = ");
        if (expr == null) {
            if (array) stringBuffer.append("{}");
            else stringBuffer.append("nil");
        } else {
            expr.accept(this);
        }
        appendStatementLineEnd();
    }

    // --- function
    @Override
    public void appendFunction(@Nullable String returns, String name, @NotNull List<JassParam> params, @NotNull List<JassStmt> statements) {
        for (JassParam param : params) {
            stringBuffer.append("---@param ").append(param.getId().getText()).append(" ").append(getConvertedTypeName(param.getTypeName().getText())).append("\n");
        }
        if (returns != null) stringBuffer.append("---@return ").append(returns).append("\n");

        stringBuffer.append("function ").append(name).append("(");
        for (int i = 0; i < params.size(); i++) {
            stringBuffer.append(params.get(i).getId().getText());
            if (i < params.size() - 1) stringBuffer.append(", ");
        }
        stringBuffer.append(")\n");
        for (JassStmt stmt : statements) stmt.accept(this);
        stringBuffer.append("end\n");
    }

    // --- statement

    // if
    @Override
    public void visitIfStmt(@NotNull JassIfStmt o) {
        stringBuffer.append("if ");
        acceptExpr(o.getExpr());
        stringBuffer.append(" then\n");
        for (JassStmt stmt : o.getStmtList()) stmt.accept(this);
        for (JassElseIfStmt stmt : o.getElseIfStmtList()) stmt.accept(this);
        for (JassElseStmt stmt : o.getElseStmtList()) stmt.accept(this);
        stringBuffer.append("end\n");
    }

    @Override
    public void visitElseIfStmt(@NotNull JassElseIfStmt o) {
        stringBuffer.append("elseif ");
        acceptExpr(o.getExpr());
        stringBuffer.append(" then\n");
        for (JassStmt stmt : o.getStmtList()) stmt.accept(this);
    }

    @Override
    public void visitElseStmt(@NotNull JassElseStmt o) {
        stringBuffer.append("else\n");
        for (JassStmt stmt : o.getStmtList()) stmt.accept(this);
    }

    // loop
    @Override
    public void visitLoopStmt(@NotNull JassLoopStmt o) {
        stringBuffer.append("while (true) do\n");
        for (JassStmt stmt : o.getStmtList()) stmt.accept(this);
        stringBuffer.append("end\n");
    }

    @Override
    public void visitExitWhenStmt(@NotNull JassExitWhenStmt o) {
        final var expr = o.getExpr();
        if (expr == null) return;
        stringBuffer.append("if ");
        expr.accept(this);
        stringBuffer.append(" then break end\n");
    }

    // --- expression

    @Override
    public void visitNeqExpr(@NotNull JassNeqExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "~=");
    }

    @Override
    public void visitAndExpr(@NotNull JassAndExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "and");
    }

    @Override
    public void visitOrExpr(@NotNull JassOrExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "or");
    }

    @Override
    public void visitNotExpr(@NotNull JassNotExpr o) {
        appendExprWithPrefixOp(o.getExpr(), "not");
    }

    @Override
    public void visitPlusExpr(@NotNull JassPlusExpr o) {
        final var list = o.getExprList();
        for (JassExpr expr : list) {
            final var c = expr.getFirstChild();
            if (c.getNode().getElementType() == STRVAL) {
                appendExprListConcatByOperator(list, "..");
                return;
            }
        }
        appendExprListConcatByOperator(list, "+");
    }
}

File: src/main/kotlin/raft/war/language/jass/openapi/actionSystem/convert/jass2lua/Jass2LuaVisitor.kt
package raft.war.language.jass.openapi.actionSystem.convert.jass2lua

import raft.war.language.jass.openapi.actionSystem.convert.Jass2AnyVisitor
import raft.war.language.jass.psi.*

class Jass2LuaVisitor internal constructor() : Jass2AnyVisitor() {
    init {
        //https://www.lua.org/manual/5.1/manual.html
        keywords = object : HashSet<String>() {
            init {
                add("break")
                add("do")
                add("end")
                add("for")
                add("in")
                add("nil")
                add("repeat")
                add("until")
                add("while")
            }
        }
    }

    override fun appendStatementLineEnd() {
        stringBuffer.append("\n")
    }

    override fun appendSingleLineComment(comment: String?) {
        stringBuffer.append("--").append(comment).append("\n")
    }

    override fun appendBlockLineComment(comment: String?) {
        stringBuffer.append("--[[").append(comment).append("--]]")
    }

    override fun appendNull() {
        stringBuffer.append("nil")
    }

    // -- typedef
    override fun appendTypeDef(name: String, base: String) {
        stringBuffer.append("---@class ").append(name).append(":").append(base)
        appendStatementLineEnd()
    }

    // -- variable
    override fun appendVar(
        constant: Boolean,
        global: Boolean,
        array: Boolean,
        type: String,
        name: String?,
        expr: raft.war.language.jass.psi.JassExpr?
    ) {
        stringBuffer.append("---@type ").append(type)
        if (global) stringBuffer.append(" @global")
        if (constant) stringBuffer.append(" constant")
        stringBuffer.append("\n")

        if (!global) stringBuffer.append("local ")
        stringBuffer.append(name)

        stringBuffer.append(" = ")
        expr?.accept(this) ?: if (array) stringBuffer.append("{}")
        else stringBuffer.append("nil")
        appendStatementLineEnd()
    }

    // --- function
    override fun appendFunction(
        returns: String?,
        name: String?,
        params: List<raft.war.language.jass.psi.JassParam?>,
        statements: List<raft.war.language.jass.psi.JassStmt?>
    ) {
        for (param in params) {
            stringBuffer.append("---@param ")
            appendSafeName(param!!.id.text)
            stringBuffer.append(" ").append(getConvertedTypeName(param.typeName.text)).append("\n")
        }
        if (returns != null) stringBuffer.append("---@return ").append(returns).append("\n")

        stringBuffer.append("function ").append(name).append("(")
        for (i in params.indices) {
            appendSafeName(params[i]!!.id.text)
            if (i < params.size - 1) stringBuffer.append(", ")
        }
        stringBuffer.append(")\n")
        for (stmt in statements) stmt!!.accept(this)
        stringBuffer.append("end\n")
    }

    // --- statement
    // if
    override fun visitIfStmt(o: raft.war.language.jass.psi.JassIfStmt) {
        stringBuffer.append("if ")
        acceptExpr(o.expr)
        stringBuffer.append(" then\n")
        for (stmt in o.stmtList) stmt.accept(this)
        for (stmt in o.elseIfStmtList) stmt.accept(this)
        for (stmt in o.elseStmtList) stmt.accept(this)
        stringBuffer.append("end\n")
    }

    override fun visitElseIfStmt(o: raft.war.language.jass.psi.JassElseIfStmt) {
        stringBuffer.append("elseif ")
        acceptExpr(o.expr)
        stringBuffer.append(" then\n")
        for (stmt in o.stmtList) stmt.accept(this)
    }

    override fun visitElseStmt(o: raft.war.language.jass.psi.JassElseStmt) {
        stringBuffer.append("else\n")
        for (stmt in o.stmtList) stmt.accept(this)
    }

    // loop
    override fun visitLoopStmt(o: raft.war.language.jass.psi.JassLoopStmt) {
        stringBuffer.append("while (true) do\n")
        for (stmt in o.stmtList) stmt.accept(this)
        stringBuffer.append("end\n")
    }

    override fun visitExitWhenStmt(o: raft.war.language.jass.psi.JassExitWhenStmt) {
        val expr = o.expr ?: return
        stringBuffer.append("if ")
        expr.accept(this)
        stringBuffer.append(" then break end\n")
    }

    // --- expression
    override fun visitNeqExpr(o: raft.war.language.jass.psi.JassNeqExpr) {
        appendExprListConcatByOperator(o.exprList, "~=")
    }

    override fun visitAndExpr(o: raft.war.language.jass.psi.JassAndExpr) {
        appendExprListConcatByOperator(o.exprList, "and")
    }

    override fun visitOrExpr(o: raft.war.language.jass.psi.JassOrExpr) {
        appendExprListConcatByOperator(o.exprList, "or")
    }

    override fun visitNotExpr(o: raft.war.language.jass.psi.JassNotExpr) {
        appendExprWithPrefixOp(o.expr, "not")
    }

    override fun visitPlusExpr(o: raft.war.language.jass.psi.JassPlusExpr) {
        val list = o.exprList
        for (expr in list) {
            val c = expr.firstChild
            if (c.node.elementType === raft.war.language.jass.psi.JassTypes.STRVAL) {
                appendExprListConcatByOperator(list, "..")
                return
            }
        }
        appendExprListConcatByOperator(list, "+")
    }
}
